drivers/src/pmu.rs

```diff
@@ -0,0 +1,48 @@
+use arrayvec::ArrayVec;
+use s_mode_utils::ecall::ecall_send;
+use sbi::PmuFunction::GetNumCounters;
+use sbi::*;
+use spin::Once;
+
+const MAX_COUNTERS: usize = 32;
+
+/// Caches information about platform hardware and fimrware counters.
+#[derive(Debug)]
+pub struct PmuInfo {
+    /// Platform counter information and the original counter index.
+    pub counters_info: ArrayVec<(PmuCounterInfo, bool), MAX_COUNTERS>,
+}
+
+static PMU_INFO: Once<PmuInfo> = Once::new();
+
+impl PmuInfo {
+    /// Initializes the global PmuInfo structure with the information obtained from the platform SBI.
+    pub fn init() {
+        let msg = SbiMessage::Pmu(GetNumCounters);
+        // Safety: PmuFunction does not touch memory.
+        let result = unsafe { ecall_send(&msg) };
```

Comment:
if you check what I did the the TSM api, it would be good to match that for PMU. So users don't need to have unsafe blocks for operations which can be wrapped in a safe helper.

Reply:
Changed.